### ISSUE TITLE
Compatibility with Socket.io

### ISSUE THREAD

**Issue by Clansty** [davidje13/websocket-express#3]
When I use websocket-express with [Socket.io](https://socket.io), the websocket endpoint of socket.io cannot be connected and returned `RSV1 must be clear`. I tried the following methods to use them together:

```ts
  const server = app.listen(7878, () => log.info("Started"));
  await initSocketIo(server)
```

```ts
  const server = createServer();
  await initSocketIo(server);
  app.attach(server);
  server.listen(7878);
```

```ts
export const initSocketIo = async (server: HttpServer) => {
  io = new SocketIoServer(server);
};
```

If I only attach socket.io, the websocket endpoint works properly. But If I also attach websocket-express, socket.io won't be connected via websocket.

I think it's websocket-express handled websocket connections on all endpoints and conflict with what socket.io do

**Comment by Clansty** [comment]
Besides `RSV1 must be clear` I also got 404 if I remove `app.use("/", express.static("./front"));`

**Comment by davidje13** [comment]
Yes websocket-express will capture _all_ websocket requests that reach the server (if it didn't, then the common use-case would suffer from long-lived connections which eventually time out whenever a user attempts to connect to a websocket at an unknown URL).

I haven't used socket.io before, but the _intent_ of websocket-express is that it handles the routing and delegates the actual management of the websocket to the handlers, which would look something like this (pseudo-code):

```js
const app = new WebSocketExpress();
const socketIORouter = new Router();

socketIORouter.ws('/my-socketio-path', async (req, res) => {
  const ws = await res.accept();
  socketIOLibrary.doSomethingWithWebsocket(ws); // maybe a function like this exists in socket.io?
});

app.use(socketIORouter);
app.listen(7878, () => log.info("Started"));
```

(of course if you don't have any non-socket.io endpoints, there's no point using websocket-express at all, but I'm assuming in your real application you do have some bespoke websocket endpoints which you're also registering)

But if you can't do that (perhaps because socket.io needs more global control over the connections), you could always register a listener which explicitly _doesn't_ handle the requests that should go to websocket.io:

```js
const app = new WebSocketExpress();
const socketIORouter = new Router();

socketIORouter.ws('/my-socketio-path', (req, res) => { // this path could be a wildcard if you want all subpaths of a folder to be socket.io connections
  // do nothing (handled by socket.io)
});

app.use(socketIORouter);
const server = app.listen(7878, () => log.info("Started"));
await initSocketIo(server);
```

Note that because it never calls either `req.accept` nor `next`, that handler will just sit on the connection but do nothing with it (the library is waiting for the code to decide whether to accept or reject the connection). This should leave the socket available for socket.io to handle however it likes.

With this approach, you might have some issues when the server shuts down: any still-open connections using this method will get a HTTP 500 error "server shutdown", which the client won't understand if the connection has already been upgraded to a websocket connection by socket.io. The socket will also close immediately, so if socket.io would normally perform some kind of graceful shutdown, make sure you trigger that _first_ before calling `server.close()`.

You might also find you still have the reverse problem: socket.io trying to handle connections which you want to use websocket-express for. If that's the case, you'd have to find some similar setup for socket.io.

Overall I'd recommend the first approach if at all possible, since it keeps everything organised in the way websocket-express expects, but the second approach is available as an escape hatch.

**Comment by Clansty** [comment]
I didn't find any functions in socket.io to directly accept ws connections, so I added `app.ws("/socket.io", () => {});`

Then I got `socket hang up` when connecting

**Comment by davidje13** [comment]
I had a play with socket.io & express, and it seems that express will change the request's URL property in a way which breaks socket.io. Fortunately it's pretty easy to work around this quirky behaviour: instead of fully ignoring the socket.io requests, you can patch them:

```js
socketIORouter.ws('/socket.io/*', (req) => {
  req.url = req.originalUrl; // undo modification of request object done by express
  // do nothing more - socket.io will handle the request
});
```

I've tested this solution locally and it appears to work fine. This is also [documented behaviour of express](https://expressjs.com/en/api.html#req.originalUrl) so presumably can be relied on to stay stable in future versions.

**Comment by davidje13** [comment]
I added a new `abandon` method which wraps this fix up, and also makes the close behaviour slightly nicer in this situation (delegating all `close` responsibilities for abandoned connections to the other library). It's available in version 3.1.0 and you can use it like this:

```js
myApp.ws('/socket.io/*', (req, res) => res.abandon());
```

**Comment by davidje13** [comment]
I'm going to assume this is resolved and close this issue now, but if you're still having problems please feel free to re-open.
